contracts/EntryPoint.sol

```diff
@@ -83,18 +94,81 @@ contract EntryPoint is StakeManager {
         require(success);
     }
 
+    //per aggregator information
+    struct AggregatorInfo {
+        // aggregator address
+        IAggregator aggregator;
+        // count of UserOps for this aggregator
+        uint count;
+        // aggregated signature
+        bytes signature;
+    }
+
+    //run all aggregators, each on its respective wallets
+    function validateAggregatedSignatures(
+        UserOperation[] calldata ops,
+        AggregatorInfo[] memory aggregators
+    ) private view returns (uint totalCount) {
+        uint startPtr = getMemoryPointer();
+
+        uint index = 0;
+        for (uint i = 0; i < aggregators.length; i++) {
+            AggregatorInfo memory agg = aggregators[i];
+            uint count = agg.count;
+            UserOperation[] memory copyOfOps = copyOpsRange(ops, index, count);
+            // solhint-disable-next-line no-empty-blocks
+            try agg.aggregator.validateSignatures(copyOfOps, agg.signature) {}
+            catch {
+                revert SignatureValidationFailed(address(agg.aggregator));
+            }
+            index += count;
+        }
+        restoreMemoryPointer(startPtr);
+        return index;
+    }
     /**
      * Execute a batch of UserOperation.
      * @param ops the operations to execute
      * @param beneficiary the address to receive the fees
+     * @param aggregators list of aggregators to validate signatures. each aggregator define a count it validates from userOps
+     * @param execOrder execution order of userOps. either an empty array, or an array of all userOps indices in the correct execution order
      */
-    function handleOps(UserOperation[] calldata ops, address payable beneficiary) public {
+    function handleOps(
+        UserOperation[] calldata ops,
+        address payable beneficiary,
+        AggregatorInfo[] calldata aggregators,
+        uint[] memory execOrder
```

Comment:
Just a question, what is our current calldata compression situation?
Because, if I remember correctly, say, this `uint[] memory` will need 32 bytes per UserOp just to tell in what order to run those operations

Reply:
That's true. RPC doesn't compress in calldata (only in storage), so even making this uint8[] wouldn't make a difference.
we _could_ make it "bytes", and manually parse it byte by byte, but I'm not sure it worth the trouble

Reply:
interesting, though 31 seems very small to me. We do need to calculate the overhead of supporting multiple aggregators (and reordering) to understand if it worth the effort.
Its true that in a world (network) where there is a single (dominant) aggregator , there is no need for neither aggregator array, nor reordering.

Reply:
compression is something that was left out, and can be handled by a separate, wrapper call to EntryPoint. need to check the benefit of such contract, as both Arbitrum and Optimism already employ generic compression on calldata.